// mpl-candy-machine/dist/test/initializeCandyMachine.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const mpl_toolbox_1 = require("@metaplex-foundation/mpl-toolbox");
const mpl_token_metadata_1 = require("@metaplex-foundation/mpl-token-metadata");
const umi_1 = require("@metaplex-foundation/umi");
const ava_1 = __importDefault(require("ava"));
const src_1 = require("../src");
const _setup_1 = require("./_setup");
/**
 * Note that most of the tests for the "initializeCandyMachine" instructions are
 * part of the "createCandyMachine" tests as they are more convenient to test.
 */
(0, ava_1.default)('it can initialize a new candy machine account', async (t) => {
    // Given an empty candy machine account with a big enough size.
    const umi = await (0, _setup_1.createUmi)();
    const candyMachine = (0, umi_1.generateSigner)(umi);
    await (0, umi_1.transactionBuilder)()
        .add((0, mpl_toolbox_1.createAccountWithRent)(umi, {
        newAccount: candyMachine,
        space: 5000,
        programId: umi.programs.get('mplCandyMachineCore').publicKey,
    }))
        .sendAndConfirm(umi);
    // And a collection NFT.
    const collectionMint = await (0, _setup_1.createCollectionNft)(umi);
    // When we initialize a candy machine at this address.
    const creator = (0, umi_1.generateSigner)(umi);
    await (0, umi_1.transactionBuilder)()
        .add((0, src_1.initializeCandyMachine)(umi, {
        candyMachine: candyMachine.publicKey,
        collectionMint: collectionMint.publicKey,
        collectionUpdateAuthority: umi.identity,
        itemsAvailable: 100,
        sellerFeeBasisPoints: (0, umi_1.percentAmount)(1.23),
        creators: [
            { address: creator.publicKey, verified: false, percentageShare: 100 },
        ],
        configLineSettings: (0, umi_1.some)({
            prefixName: 'My NFT #',
            nameLength: 8,
            prefixUri: 'https://example.com/',
            uriLength: 20,
            isSequential: false,
        }),
    }))
        .sendAndConfirm(umi);
    // Then we expect the candy machine account to have the right data.
    const candyMachineAccount = await (0, src_1.fetchCandyMachine)(umi, candyMachine.publicKey);
    t.like(candyMachineAccount, {
        publicKey: (0, umi_1.publicKey)(candyMachine),
        authority: (0, umi_1.publicKey)(umi.identity),
        mintAuthority: (0, umi_1.publicKey)(umi.identity),
        collectionMint: (0, umi_1.publicKey)(collectionMint),
        version: src_1.AccountVersion.V1,
        tokenStandard: mpl_token_metadata_1.TokenStandard.NonFungible,
        itemsRedeemed: 0n,
        data: {
            itemsAvailable: 100n,
            symbol: '',
            sellerFeeBasisPoints: (0, umi_1.percentAmount)(1.23),
            maxEditionSupply: 0n,
            isMutable: true,
            creators: [
                {
                    address: (0, umi_1.publicKey)(creator),
                    verified: false,
                    percentageShare: 100,
                },
            ],
            configLineSettings: (0, umi_1.some)({
                prefixName: 'My NFT #',
                nameLength: 8,
                prefixUri: 'https://example.com/',
                uriLength: 20,
                isSequential: false,
            }),
            hiddenSettings: (0, umi_1.none)(),
        },
    });
});
