// mpl-candy-machine/dist/test/setCollection.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const umi_1 = require("@metaplex-foundation/umi");
const ava_1 = __importDefault(require("ava"));
const src_1 = require("../src");
const _setup_1 = require("./_setup");
(0, ava_1.default)('it can update the collection of a candy machine v1', async (t) => {
    // Given a Candy Machine associated with Collection A.
    const umi = await (0, _setup_1.createUmi)();
    const collectionUpdateAuthorityA = (0, umi_1.generateSigner)(umi);
    const collectionA = await (0, _setup_1.createCollectionNft)(umi, {
        authority: collectionUpdateAuthorityA,
    });
    const candyMachine = await (0, _setup_1.createV1)(umi, {
        collectionMint: collectionA.publicKey,
        collectionUpdateAuthority: collectionUpdateAuthorityA,
    });
    // When we update its collection to Collection B.
    const collectionUpdateAuthorityB = (0, umi_1.generateSigner)(umi);
    const collectionB = await (0, _setup_1.createCollectionNft)(umi, {
        authority: collectionUpdateAuthorityB,
    });
    await (0, umi_1.transactionBuilder)()
        .add((0, src_1.setCollection)(umi, {
        candyMachine: candyMachine.publicKey,
        collectionMint: collectionA.publicKey,
        newCollectionMint: collectionB.publicKey,
        newCollectionUpdateAuthority: collectionUpdateAuthorityB,
    }))
        .sendAndConfirm(umi);
    // Then the Candy Machine's collection was updated accordingly.
    const candyMachineAccount = await (0, src_1.fetchCandyMachine)(umi, candyMachine.publicKey);
    t.like(candyMachineAccount, {
        collectionMint: (0, umi_1.publicKey)(collectionB.publicKey),
    });
});
(0, ava_1.default)('it cannot update the collection of a candy machine v2', async (t) => {
    // Given a Candy Machine v2 associated with Collection A.
    const umi = await (0, _setup_1.createUmi)();
    const collectionUpdateAuthorityA = (0, umi_1.generateSigner)(umi);
    const collectionA = await (0, _setup_1.createCollectionNft)(umi, {
        authority: collectionUpdateAuthorityA,
    });
    const candyMachine = await (0, _setup_1.createV2)(umi, {
        collectionMint: collectionA.publicKey,
        collectionUpdateAuthority: collectionUpdateAuthorityA,
    });
    // When we try to update its collection using the setCollection v1 instruction.
    const collectionUpdateAuthorityB = (0, umi_1.generateSigner)(umi);
    const collectionB = await (0, _setup_1.createCollectionNft)(umi, {
        authority: collectionUpdateAuthorityB,
    });
    const promise = (0, umi_1.transactionBuilder)()
        .add((0, src_1.setCollection)(umi, {
        candyMachine: candyMachine.publicKey,
        collectionMint: collectionA.publicKey,
        newCollectionMint: collectionB.publicKey,
        newCollectionUpdateAuthority: collectionUpdateAuthorityB,
    }))
        .sendAndConfirm(umi);
    // Then we expect a program error.
    await t.throwsAsync(promise, { message: /Use SetCollectionV2 instead/ });
});
