// mpl-candy-machine/dist/test/defaultGuards/mintLimit.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const mpl_toolbox_1 = require("@metaplex-foundation/mpl-toolbox");
const umi_1 = require("@metaplex-foundation/umi");
const ava_1 = __importDefault(require("ava"));
const src_1 = require("../../src");
const _setup_1 = require("../_setup");
(0, ava_1.default)('it allows minting when the mint limit is not reached', async (t) => {
    // Given a loaded Candy Machine with a mint limit of 5.
    const umi = await (0, _setup_1.createUmi)();
    const collectionMint = (await (0, _setup_1.createCollectionNft)(umi)).publicKey;
    const { publicKey: candyMachine } = await (0, _setup_1.createV2)(umi, {
        collectionMint,
        configLines: [
            { name: 'Degen #1', uri: 'https://example.com/degen/1' },
            { name: 'Degen #2', uri: 'https://example.com/degen/2' },
        ],
        guards: {
            mintLimit: (0, umi_1.some)({ id: 1, limit: 5 }),
        },
    });
    // When we mint from it.
    const mint = (0, umi_1.generateSigner)(umi);
    await (0, umi_1.transactionBuilder)()
        .add((0, mpl_toolbox_1.setComputeUnitLimit)(umi, { units: 600000 }))
        .add((0, src_1.mintV2)(umi, {
        candyMachine,
        nftMint: mint,
        collectionMint,
        collectionUpdateAuthority: umi.identity.publicKey,
        mintArgs: { mintLimit: (0, umi_1.some)({ id: 1 }) },
    }))
        .sendAndConfirm(umi);
    // Then minting was successful.
    await (0, _setup_1.assertSuccessfulMint)(t, umi, { mint, owner: umi.identity });
    // And the mint limit PDA was incremented.
    const counterPda = (0, src_1.findMintCounterPda)(umi, {
        id: 1,
        user: umi.identity.publicKey,
        candyMachine,
        candyGuard: (0, src_1.findCandyGuardPda)(umi, { base: candyMachine })[0],
    });
    const counterAccount = await (0, src_1.fetchMintCounter)(umi, counterPda);
    t.is(counterAccount.count, 1);
});
(0, ava_1.default)('it allows minting even when the payer is different from the minter', async (t) => {
    // Given a loaded Candy Machine with a mint limit of 5.
    const umi = await (0, _setup_1.createUmi)();
    const collectionMint = (await (0, _setup_1.createCollectionNft)(umi)).publicKey;
    const { publicKey: candyMachine } = await (0, _setup_1.createV2)(umi, {
        collectionMint,
        configLines: [
            { name: 'Degen #1', uri: 'https://example.com/degen/1' },
            { name: 'Degen #2', uri: 'https://example.com/degen/2' },
        ],
        guards: {
            mintLimit: (0, umi_1.some)({ id: 1, limit: 5 }),
        },
    });
    // When we mint from it using a separate minter.
    const minter = (0, umi_1.generateSigner)(umi);
    const mint = (0, umi_1.generateSigner)(umi);
    await (0, umi_1.transactionBuilder)()
        .add((0, mpl_toolbox_1.setComputeUnitLimit)(umi, { units: 600000 }))
        .add((0, src_1.mintV2)(umi, {
        candyMachine,
        nftMint: mint,
        minter,
        collectionMint,
        collectionUpdateAuthority: umi.identity.publicKey,
        mintArgs: { mintLimit: (0, umi_1.some)({ id: 1 }) },
    }))
        .sendAndConfirm(umi);
    // Then minting was successful.
    await (0, _setup_1.assertSuccessfulMint)(t, umi, { mint, owner: minter });
    // And the mint limit PDA was incremented for that minter.
    const counterPda = (0, src_1.findMintCounterPda)(umi, {
        id: 1,
        user: minter.publicKey,
        candyMachine,
        candyGuard: (0, src_1.findCandyGuardPda)(umi, { base: candyMachine })[0],
    });
    const counterAccount = await (0, src_1.fetchMintCounter)(umi, counterPda);
    t.is(counterAccount.count, 1);
});
(0, ava_1.default)('it forbids minting when the mint limit is reached', async (t) => {
    // Given a loaded Candy Machine with a mint limit of 1.
    const umi = await (0, _setup_1.createUmi)();
    const collectionMint = (await (0, _setup_1.createCollectionNft)(umi)).publicKey;
    const { publicKey: candyMachine } = await (0, _setup_1.createV2)(umi, {
        collectionMint,
        configLines: [
            { name: 'Degen #1', uri: 'https://example.com/degen/1' },
            { name: 'Degen #2', uri: 'https://example.com/degen/2' },
        ],
        guards: {
            mintLimit: (0, umi_1.some)({ id: 42, limit: 1 }),
        },
    });
    // And the identity already minted their NFT.
    const mint = (0, umi_1.generateSigner)(umi);
    await (0, umi_1.transactionBuilder)()
        .add((0, mpl_toolbox_1.setComputeUnitLimit)(umi, { units: 600000 }))
        .add((0, src_1.mintV2)(umi, {
        candyMachine,
        nftMint: mint,
        collectionMint,
        collectionUpdateAuthority: umi.identity.publicKey,
        mintArgs: { mintLimit: (0, umi_1.some)({ id: 42 }) },
    }))
        .sendAndConfirm(umi);
    // When that same identity tries to mint from the same Candy Machine again.
    const promise = (0, umi_1.transactionBuilder)()
        .add((0, mpl_toolbox_1.setComputeUnitLimit)(umi, { units: 600000 }))
        .add((0, src_1.mintV2)(umi, {
        candyMachine,
        nftMint: (0, umi_1.generateSigner)(umi),
        collectionMint,
        collectionUpdateAuthority: umi.identity.publicKey,
        mintArgs: { mintLimit: (0, umi_1.some)({ id: 42 }) },
    }))
        .sendAndConfirm(umi);
    // Then we expect an error.
    await t.throwsAsync(promise, { message: /AllowedMintLimitReached/ });
});
(0, ava_1.default)('the mint limit is local to each wallet', async (t) => {
    // Given a loaded Candy Machine with a mint limit of 1.
    const umi = await (0, _setup_1.createUmi)();
    const collectionMint = (await (0, _setup_1.createCollectionNft)(umi)).publicKey;
    const { publicKey: candyMachine } = await (0, _setup_1.createV2)(umi, {
        collectionMint,
        configLines: [
            { name: 'Degen #1', uri: 'https://example.com/degen/1' },
            { name: 'Degen #2', uri: 'https://example.com/degen/2' },
        ],
        guards: {
            mintLimit: (0, umi_1.some)({ id: 42, limit: 1 }),
        },
    });
    // And minter A already minted their NFT.
    const minterA = (0, umi_1.generateSigner)(umi);
    const mintA = (0, umi_1.generateSigner)(umi);
    await (0, umi_1.transactionBuilder)()
        .add((0, mpl_toolbox_1.setComputeUnitLimit)(umi, { units: 600000 }))
        .add((0, src_1.mintV2)(umi, {
        candyMachine,
        nftMint: mintA,
        minter: minterA,
        collectionMint,
        collectionUpdateAuthority: umi.identity.publicKey,
        mintArgs: { mintLimit: (0, umi_1.some)({ id: 42 }) },
    }))
        .sendAndConfirm(umi);
    await (0, _setup_1.assertSuccessfulMint)(t, umi, { mint: mintA, owner: minterA });
    // When minter B mints from the same Candy Machine.
    const minterB = (0, umi_1.generateSigner)(umi);
    const mintB = (0, umi_1.generateSigner)(umi);
    await (0, umi_1.transactionBuilder)()
        .add((0, mpl_toolbox_1.setComputeUnitLimit)(umi, { units: 600000 }))
        .add((0, src_1.mintV2)(umi, {
        candyMachine,
        nftMint: mintB,
        minter: minterB,
        collectionMint,
        collectionUpdateAuthority: umi.identity.publicKey,
        mintArgs: { mintLimit: (0, umi_1.some)({ id: 42 }) },
    }))
        .sendAndConfirm(umi);
    // Then minting was successful as the limit is per wallet.
    await (0, _setup_1.assertSuccessfulMint)(t, umi, { mint: mintB, owner: minterB });
});
(0, ava_1.default)('it charges a bot tax when trying to mint after the limit', async (t) => {
    // Given a loaded Candy Machine with a mint limit of 1 and a bot tax guard.
    const umi = await (0, _setup_1.createUmi)();
    const collectionMint = (await (0, _setup_1.createCollectionNft)(umi)).publicKey;
    const { publicKey: candyMachine } = await (0, _setup_1.createV2)(umi, {
        collectionMint,
        configLines: [
            { name: 'Degen #1', uri: 'https://example.com/degen/1' },
            { name: 'Degen #2', uri: 'https://example.com/degen/2' },
        ],
        guards: {
            botTax: (0, umi_1.some)({ lamports: (0, umi_1.sol)(0.1), lastInstruction: true }),
            mintLimit: (0, umi_1.some)({ id: 42, limit: 1 }),
        },
    });
    // And the identity already minted their NFT.
    const mintA = (0, umi_1.generateSigner)(umi);
    await (0, umi_1.transactionBuilder)()
        .add((0, mpl_toolbox_1.setComputeUnitLimit)(umi, { units: 600000 }))
        .add((0, src_1.mintV2)(umi, {
        candyMachine,
        nftMint: mintA,
        collectionMint,
        collectionUpdateAuthority: umi.identity.publicKey,
        mintArgs: { mintLimit: (0, umi_1.some)({ id: 42 }) },
    }))
        .sendAndConfirm(umi);
    // When the identity tries to mint from the same Candy Machine again.
    const mintB = (0, umi_1.generateSigner)(umi);
    const { signature } = await (0, umi_1.transactionBuilder)()
        .add((0, mpl_toolbox_1.setComputeUnitLimit)(umi, { units: 600000 }))
        .add((0, src_1.mintV2)(umi, {
        candyMachine,
        nftMint: mintB,
        collectionMint,
        collectionUpdateAuthority: umi.identity.publicKey,
        mintArgs: { mintLimit: (0, umi_1.some)({ id: 42 }) },
    }))
        .sendAndConfirm(umi);
    // Then we expect a bot tax error.
    await (0, _setup_1.assertBotTax)(t, umi, mintB, signature, /AllowedMintLimitReached/);
});
