// mpl-candy-machine/dist/test/defaultGuards/thirdPartySigner.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const mpl_toolbox_1 = require("@metaplex-foundation/mpl-toolbox");
const umi_1 = require("@metaplex-foundation/umi");
const ava_1 = __importDefault(require("ava"));
const src_1 = require("../../src");
const _setup_1 = require("../_setup");
(0, ava_1.default)('it allows minting when the third party signer is provided', async (t) => {
    // Given a loaded Candy Machine with a third party signer guard.
    const umi = await (0, _setup_1.createUmi)();
    const thirdPartySigner = (0, umi_1.generateSigner)(umi);
    const collectionMint = (await (0, _setup_1.createCollectionNft)(umi)).publicKey;
    const { publicKey: candyMachine } = await (0, _setup_1.createV2)(umi, {
        collectionMint,
        configLines: [{ name: 'Degen #1', uri: 'https://example.com/degen/1' }],
        guards: {
            thirdPartySigner: (0, umi_1.some)({ signerKey: thirdPartySigner.publicKey }),
        },
    });
    // When we mint from it by providing the third party as a Signer.
    const mint = (0, umi_1.generateSigner)(umi);
    await (0, umi_1.transactionBuilder)()
        .add((0, mpl_toolbox_1.setComputeUnitLimit)(umi, { units: 600000 }))
        .add((0, src_1.mintV2)(umi, {
        candyMachine,
        nftMint: mint,
        collectionMint,
        collectionUpdateAuthority: umi.identity.publicKey,
        mintArgs: {
            thirdPartySigner: (0, umi_1.some)({ signer: thirdPartySigner }),
        },
    }))
        .sendAndConfirm(umi);
    // Then minting was successful.
    await (0, _setup_1.assertSuccessfulMint)(t, umi, { mint, owner: umi.identity });
});
(0, ava_1.default)('it forbids minting when the third party signer is wrong', async (t) => {
    // Given a loaded Candy Machine with a third party signer guard.
    const umi = await (0, _setup_1.createUmi)();
    const thirdPartySigner = (0, umi_1.generateSigner)(umi);
    const collectionMint = (await (0, _setup_1.createCollectionNft)(umi)).publicKey;
    const { publicKey: candyMachine } = await (0, _setup_1.createV2)(umi, {
        collectionMint,
        configLines: [{ name: 'Degen #1', uri: 'https://example.com/degen/1' }],
        guards: {
            thirdPartySigner: (0, umi_1.some)({ signerKey: thirdPartySigner.publicKey }),
        },
    });
    // When we try to mint from it by providing the wrong third party signer.
    const wrongThirdPartySigner = (0, umi_1.generateSigner)(umi);
    const mint = (0, umi_1.generateSigner)(umi);
    const promise = (0, umi_1.transactionBuilder)()
        .add((0, mpl_toolbox_1.setComputeUnitLimit)(umi, { units: 600000 }))
        .add((0, src_1.mintV2)(umi, {
        candyMachine,
        nftMint: mint,
        collectionMint,
        collectionUpdateAuthority: umi.identity.publicKey,
        mintArgs: {
            thirdPartySigner: (0, umi_1.some)({ signer: wrongThirdPartySigner }),
        },
    }))
        .sendAndConfirm(umi);
    // Then we expect an error.
    await t.throwsAsync(promise, { message: /MissingRequiredSignature/ });
});
(0, ava_1.default)('it charges a bot tax when trying to mint using the wrong third party signer', async (t) => {
    // Given a loaded Candy Machine with a third party signer guard and a bot tax guard.
    const umi = await (0, _setup_1.createUmi)();
    const thirdPartySigner = (0, umi_1.generateSigner)(umi);
    const collectionMint = (await (0, _setup_1.createCollectionNft)(umi)).publicKey;
    const { publicKey: candyMachine } = await (0, _setup_1.createV2)(umi, {
        collectionMint,
        configLines: [{ name: 'Degen #1', uri: 'https://example.com/degen/1' }],
        guards: {
            botTax: (0, umi_1.some)({ lamports: (0, umi_1.sol)(0.1), lastInstruction: true }),
            thirdPartySigner: (0, umi_1.some)({ signerKey: thirdPartySigner.publicKey }),
        },
    });
    // When we try to mint from it by providing the wrong third party signer.
    const wrongThirdPartySigner = (0, umi_1.generateSigner)(umi);
    const mint = (0, umi_1.generateSigner)(umi);
    const { signature } = await (0, umi_1.transactionBuilder)()
        .add((0, mpl_toolbox_1.setComputeUnitLimit)(umi, { units: 600000 }))
        .add((0, src_1.mintV2)(umi, {
        candyMachine,
        nftMint: mint,
        collectionMint,
        collectionUpdateAuthority: umi.identity.publicKey,
        mintArgs: {
            thirdPartySigner: (0, umi_1.some)({ signer: wrongThirdPartySigner }),
        },
    }))
        .sendAndConfirm(umi);
    // Then we expect a silent bot tax error.
    await (0, _setup_1.assertBotTax)(t, umi, mint, signature, /MissingRequiredSignature/);
});
